enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,255 @@
+Finally the status section of the `insightsoperator.operator.openshift.io` CR is updated. Note that there is a retry logic in case of failed upload and after some time (and number of attempts) the Insights operator is marked as degraded.
+Introducing a new on demand process/container for running an "extra" gather process raises the question about parallelism. It should be feasible to allow two (or more) parallel gather processes, but it is very important to keep in mind that only one (and that is the latest) Insights analysis report will be available.  
+
+
+### Risks and Mitigations
+
+Risk: Inconsistent Insights analysis results. Assume following situation:
+- periodic gathering is running in the Insights operator (let's call it gather A)
+- user runs a new on demand data gathering (let's call it gather B) e.g 30 seconds after gather A
+- gather A and B finish at essentially the same time
+- the external processing service (in the console.redhat.com) provides only one recent Insights analysis result, and there is no guarantee which (of the two) it is.
+- this can be very confusing, because user running gather B can get result of the gather A (e.g there could be different gatherers enabled, which would produce different results)
+
+Mitigation: This is the very first question in the [Open questions](#open-questions) section, but perhaps reporting the Insights analysis result in one place could slightly mitigate this problem (the status part could tell whether it was a periodic gather or user-triggered gather). 
+Another option would be to disallow the parallel gather execution - if the Insights operator sees a new CR instance (or any other running gather process) then it interrupts any gathering in progress. 
+
+### Drawbacks
+
+- Very frequent data gathering in a cluster can produce a significant load on the processing part in the console.redhat.com and can overwhelm the processing pipeline. 
+Theoretically, we could limit the number of user-trigerred gathering over time, but this problem can happen today when something is deleting the Insights operator pod on a regular basis with appropriate timing (so that data can be uploaded).  
+- When the newly spawned container is failing to upload the data then it can be running for some longer time (couple of hours). 
+
+
+## Design Details
+
+### Open Questions
+- How to solve the problem with only one Insights analysis result available in a cluster? How to avoid the "uncertainty" of the Insights analysis results described in the [Risk and Mitigations](#risks-and-mitigations) section? 
+- Should the newly created gathering process update the clusteroperator conditions (e.g it fails to upload the data, should it update the `Degraded` status as it is today)? This can probably be racy too or it can lead to temporary and inconsistent conditions state. 
```

Comment:
If you had multiple containers running, this could be racy as you've said. What if the operator creating the pods observed their status instead and could then report back.

I'm imagining something along the lines of:
* Process gathers insights data and stores it in an `emptyDir`
* Process attempts upload of data and exits with an error code if this fails
* K8s restarts pod, process sees existing data, skips to upload stage, exits 0 when the upload succeeds, else loop

This would then allow the operator owning the pods to see, hey, that pod is restarting lots of times, it must have an issue, I'll report and issue to the clusteroperator